router/throttler/adaptiveAlgorithmCounter/algorithm.go

```diff
@@ -0,0 +1,102 @@
+package adaptivethrottlercounter
+
+import (
+	"context"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/rudderlabs/rudder-go-kit/config"
+)
+
+type timer struct {
+	frequency    time.Duration
+	limitReached bool
+	mu           sync.Mutex
+	limitSet     bool
+	cancel       context.CancelFunc
+}
+
+type Adaptive struct {
+	shortTimer               *timer
+	longTimer                *timer
+	decreaseLimitPercentage  *config.Reloadable[int64]
+	increaseChangePercentage *config.Reloadable[int64]
+}
+
+func New(config *config.Config) *Adaptive {
+	shortTimeFrequency := config.GetDuration("Router.throttler.adaptive.shortTimeFrequency", 5, time.Second)
+	longTimeFrequency := config.GetDuration("Router.throttler.adaptive.longTimeFrequency", 15, time.Second)
+
+	shortTimer := &timer{
+		frequency: shortTimeFrequency,
+	}
+	longTimer := &timer{
+		frequency: longTimeFrequency,
+	}
+
+	go shortTimer.run()
+	go longTimer.run()
```

Comment:
We need to make sure that these stop during shutdown. We may use an `errgroup` or a `sync.WaitGroup`

Reply:
We are canceling them during` shutdown()` right?